### PR TITLE
[NCL-5765] improved BuildStatus documentation

### PR DESCRIPTION
**Pull request by michalovjan**
### Checklist:

* [ ] Have you added a note in the [CHANGELOG wiki](https://github.com/project-ncl/pnc/wiki/Changelog) for your change if user-facing?
* [ ] Have you added unit tests for your change?

### CODE REVIEW

File: constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java [hunk 1]
```diff
@@ -24,45 +24,58 @@
...
-     * Build completed successfully.
+     * Build/GroupBuild has completed successfully. The dependant Builds can be scheduled.
      */
     SUCCESS(FINISHED, true),
 
     /**
-     * Build failed.
+     * Build/GroupBuild has failed. The status is propagated through dependants as REJECTED_FAILED_DEPENDENCIES.
+     *
+     * A GroupBuild is reported FAILED with is at least one underlying Build with the final status of (FAILED, REJECTED,
```
Comment (janinko) [inline constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java:43]: `s/with is/when/` `s/with the/has the/`? Or smething like that, it doesn't read well now.

File: constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java [hunk 2]
```diff
@@ -24,45 +24,58 @@
...
+     * SYSTEM_ERROR)
      */
     FAILED(FINISHED, false),
 
     /**
-     * A build has been requested (possibly via dependencies) but no actual build happened as it was not required (no
-     * updates).
+     * A Build has been requested (possibly via dependencies) but no actual build happened as it was not required (no
```
Comment (janinko) [inline constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java:49]: dependencies? I think it's dependants.
But do we require new build of a dependency when it doesn't require rebuild? Why and in which cases?

File: constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java [hunk 3]
```diff
@@ -24,45 +24,58 @@
...
+     * updates to underlying BuildConfig and Configs of dependencies (depends in IMPLICIT/EXPLICIT dependency check)).
+     * 
+     * A GroupBuild is deemed unnecessary to be built if all the underlying Builds do not require rebuild.
```
Comment (janinko) [inline constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java:52]: `s/underlying/included/` or something like that?

File: constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java [hunk 4]
```diff
@@ -24,45 +24,58 @@
...
+     * A GroupBuild is reported FAILED with is at least one underlying Build with the final status of (FAILED, REJECTED,
+     * SYSTEM_ERROR)
      */
     FAILED(FINISHED, false),
 
     /**
-     * A build has been requested (possibly via dependencies) but no actual build happened as it was not required (no
-     * updates).
+     * A Build has been requested (possibly via dependencies) but no actual build happened as it was not required (no
+     * updates to underlying BuildConfig and Configs of dependencies (depends in IMPLICIT/EXPLICIT dependency check)).
```
Comment (janinko) [inline constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java:50]: I would make it more clear by "(rebuild is required when there is ..."

I think rebuild is required when:
 * the BC changed
 * an explicit/implicit (transitive) dependency BC was rebuilt
 * an explicit/implicit (transitive) dependency BC changed
(if this is not true, do we have good implementation?)

File: constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java [hunk 5]
```diff
@@ -24,45 +24,58 @@
...
+     * 
+     * A GroupBuild is deemed unnecessary to be built if all the underlying Builds do not require rebuild.
      */
     NO_REBUILD_REQUIRED(FINISHED, true),
 
     /**
-     * A build has been placed in a queue. It will be processed shortly.
+     * Build satisfies all necessary requirements to start building process. Build is placed into a queue and waits for a free
```
Comment (janinko) [inline constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java:57]: Why did you remove the articles? I belive there should be `a`? @alexcreasy

File: constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java [hunk 6]
```diff
@@ -24,45 +24,58 @@
...
+     * BuildCoordinator thread to be picked up.
+     *
+     * With lower number of concurrent Builds, the Build is more likely to wait for available thread.
      */
     ENQUEUED(PENDING),
 
     /**
-     * Build is waiting for dependencies to finish.
+     * A Build has unfinished dependencies, therefore cannot start building.
```
Comment (janinko) [inline constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java:65]: Here you added article `a`

File: constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java [hunk 7]
```diff
@@ -24,45 +24,58 @@
...
      */
     WAITING_FOR_DEPENDENCIES(PENDING),
 
     /**
-     * Build currently running.
+     * The Build/GroupBuild building process has started.
```
Comment (janinko) [inline constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java:70]: and here you added article `the`. Please make your mind :D 
cc @alexcreasy what should be correct?

File: constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java (hunk 6 above)
Comment (janinko) [reply constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java:65]: I would use word waiting somewhere. Now the description looks like the build failed.

File: constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java (hunk 1 above)
Comment (michalovjan) [reply constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java:43]: +:100:

File: constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java (hunk 3 above)
Comment (michalovjan) [reply constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java:52]: +:100:

File: constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java (hunk 7 above)
Comment (michalovjan) [reply constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java:70]: In my defense 'the' applies to process

File: constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java (hunk 6 above)
Comment (michalovjan) [reply constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java:65]: +:100:

File: constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java (hunk 5 above)
Comment (michalovjan) [reply constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java:57]: Yeah, there should be 'a'.

File: constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java (hunk 2 above)
Comment (michalovjan) [reply constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java:49]: Good catch, it should be dependants. 

We require a rebuild of dependency if its BC has changed or if dependency's dependency needs rebuild(and so on). How you define which dependencies you are checking relies on IMPLICIT/EXPLICIT parameter.

File: constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java (hunk 4 above)
Comment (michalovjan) [reply constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java:50]: Sure I can make it much more clear.

File: constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java (hunk 2 above)
Comment (janinko) [reply constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java:49]: Yeah, but you say "A Build has been requested (possibly via dependants)". Do we request dependecies to build even when they don't require rebuild?

File: constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java (hunk 4 above)
Comment (matedo1) [reply constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java:50]: Regarding rebuild possibilities, there are three modes:

- **Explicit**
    - The build config has been modified since the last successful build
    - A build config, explicitly defined as a dependency of this one, has been modified
    - since the last successful build
    - There exists a newer, successful build of a build config explicitly defined as a dependency of this one
- **Implicit**
    - Explicit criteria plus:
    - There exists a newer version of an implicit dependency (automatically
    - captured from sources such as Indy, MRRC or Maven Central)
- **Force**
    - Always

Comment (alexcreasy) [reply constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java:50]: @matedo1 that's the definition we use in the UI right? Hopefully someone with a better idea of the backend implementation, e.g. Honza or Matej can verify that our description is still accurate.

File: constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java (hunk 5 above)
Comment (alexcreasy) [reply constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java:57]: Yes, the indefinite article is necessary here

File: constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java (hunk 2 above)
Comment (michalovjan) [reply constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java:49]: Yes. If you request a build, it goes through explicit dependencies. The dependency is requested, but it finishes immediately with NO_REBUILD_REQUIRED(REJECTED_ALREADY_BUILT on prod) if its BC was not modified.

File: constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java (hunk 4 above)
Comment (matedo1) [reply constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java:50]: @alexcreasy yes, the definition is coming from the UI

Comment (michalovjan) [reply constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java:50]: @alexcreasy @matedo1 WDYT about changing the Implicit description a little. For me, it's a little confusing with what we have now. 
I'd add a clause similar to this: 
     - implicit dependencies are also Builds that produced artifacts used during build process

File: constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java (hunk 2 above)
Comment (janinko) [reply constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java:49]: I don't like that behaviour, but fine :)

File: constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java (hunk 4 above)
Comment (alexcreasy) [reply constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java:50]: @michalovjan Martin's paste is a bit innacurate, it is missing the first line:

"Rebuild a previously successful build config when:"

So the bullet points are the minimum criteria which must be met for a rebuild to take place, for each rebuild mode. Your description of "implicit builds" makes no sense in that context, as it's not a criteria that must be met for a rebuild to take place.

If you want to rework the text some more so that your clarification fits in overall, by all means feel free. I don't have a complete understanding of rebuild modes, as they're exclusively backend features, so I'm not the best person to write / critique. Small modifications need to fit the overal context though.

Here's the full text from the UI popup menu, for reference:

```
Rebuild a previously successful build config when:

**Explicit**
- The build config has been modified since the last successful build
- A build config, explicitly defined as a dependency of this one, has been modified since the last successful build
- There exists a newer, successful build of a build config explicitly defined as a dependency of this one.

**Implicit**
- Explicit criteria plus:
- There exists a newer version of an implicit dependency (automatically captured from sources such as Indy, MRRC or Maven Central)

**Force**
- Always
```

Comment (alexcreasy) [reply constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java:50]: Oh on second reading, I'm wondering if you were just proposing to change the implicit clause rather than add a new one?

Comment (michalovjan) [reply constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java:50]: Yes, only implicit clause. I thought maybe instead of using parentheses, use further indentation.

Rebuild a previously successful build config when:

**Explicit**
- The build config has been modified since the last successful build
- A build config, explicitly defined as a dependency of this one, has been modified since the last successful build
- There exists a newer, successful build of a build config explicitly defined as a dependency of this one.

**Implicit**
- Explicit criteria plus:
- There exists a newer version of an implicit dependency that can be:
        - automatically captured from sources such as Indy, MRRC or Maven Central
        - a Build which produced artifacts that were used for alignment during build process

**Force**
- Always

Comment (michalovjan) [reply constants/src/main/java/org/jboss/pnc/enums/BuildStatus.java:50]: ^ @alexcreasy @matedo1

**Comment by michalovjan** [comment]
The formatting for Dependency checks is weird due to our code-formatter. It merges everything into one line unless there is empty line in between. It makes making bullet points hell.

**Comment by michalovjan** [comment]
retest this please
